controllers/clustertopology_controller_compute.go

```diff
@@ -78,8 +84,286 @@ func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster
+	// InfrastructureCluster and the ControlPlane objects generated by the previous step.
+	desiredState.cluster = computeCluster(current, desiredState.infrastructureCluster, desiredState.controlPlane.object)
+
+	// TODO: implement generate desired state for machine deployments
+
+	return desiredState, nil
+}
+
+// computeInfrastructureCluster computes the desired state for the InfrastructureCluster object starting from the
+// corresponding template defined in cluster class.
+func computeInfrastructureCluster(class *clusterTopologyClass, current *clusterTopologyState) (*unstructured.Unstructured, error) {
+	infrastructureCluster, err := templateToObject(templateToInput{
+		template:              class.infrastructureClusterTemplate,
+		templateClonedFromRef: class.clusterClass.Spec.Infrastructure.Ref,
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-", current.cluster.Name),
+		currentObjectRef:      current.cluster.Spec.InfrastructureRef,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate the cluster infrastructure object")
+	}
+	return infrastructureCluster, nil
+}
+
+// computeControlPlaneInfrastructureMachineTemplate computes the desired state for InfrastructureMachineTemplate
+// that should be referenced by the control plane object.
+func computeControlPlaneInfrastructureMachineTemplate(class *clusterTopologyClass, current *clusterTopologyState) *unstructured.Unstructured {
+	controlPlaneInfrastructureMachineTemplate := templateToTemplate(templateToInput{
+		template:              class.controlPlane.infrastructureMachineTemplate,
+		templateClonedFromRef: objToRef(class.controlPlane.infrastructureMachineTemplate),
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-controlplane-", current.cluster.Name),
+		currentObjectRef:      getNestedRef(current.controlPlane.object, "spec", "machineTemplate", "infrastructureRef", "name"),
+	})
+	return controlPlaneInfrastructureMachineTemplate
+}
+
+// computeControlPlane computes the desired state for the ControlPlane object starting from the
+// corresponding template defined in cluster class.
+func computeControlPlane(class *clusterTopologyClass, current *clusterTopologyState, infrastructureMachineTemplate *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	controlPlane, err := templateToObject(templateToInput{
+		template:              class.controlPlane.template,
+		templateClonedFromRef: class.clusterClass.Spec.ControlPlane.Ref,
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-", current.cluster.Name),
+		currentObjectRef:      current.cluster.Spec.ControlPlaneRef,
+		labels:                current.cluster.Spec.Topology.ControlPlane.Metadata.Labels,
+		annotations:           current.cluster.Spec.Topology.ControlPlane.Metadata.Annotations,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate the control plane object")
```

Comment:
```suggestion
		return nil, errors.Wrap(err, "failed to generate the ControlPlane object")
```
In case we want to align the errors to this. (would apply to a few more errors below too)